lzy/lzy-service/src/main/java/ai/lzy/service/ExecutionFinalizer.java

```diff
@@ -0,0 +1,237 @@
+package ai.lzy.service;
+
+import ai.lzy.longrunning.Operation;
+import ai.lzy.service.data.dao.ExecutionDao;
+import ai.lzy.service.data.dao.GraphDao;
+import ai.lzy.service.data.dao.PortalDescription;
+import ai.lzy.service.data.dao.WorkflowDao;
+import ai.lzy.util.auth.credentials.RenewableJwt;
+import ai.lzy.v1.AllocatorGrpc;
+import ai.lzy.v1.VmAllocatorApi;
+import ai.lzy.v1.channel.LzyChannelManagerPrivateGrpc;
+import ai.lzy.v1.graph.GraphExecutorApi;
+import ai.lzy.v1.graph.GraphExecutorGrpc;
+import ai.lzy.v1.longrunning.LongRunning;
+import ai.lzy.v1.portal.LzyPortalApi;
+import ai.lzy.v1.portal.LzyPortalGrpc;
+import com.google.common.net.HostAndPort;
+import com.google.protobuf.Any;
+import com.google.protobuf.Empty;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import jakarta.annotation.Nullable;
+import jakarta.inject.Named;
+import jakarta.inject.Singleton;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.Map;
+import java.util.Objects;
+
+import static ai.lzy.model.db.DbHelper.defaultRetryPolicy;
+import static ai.lzy.model.db.DbHelper.withRetries;
+import static ai.lzy.service.LzyService.APP;
+import static ai.lzy.util.grpc.GrpcUtils.newBlockingClient;
+import static ai.lzy.util.grpc.GrpcUtils.newGrpcChannel;
+
+@Singleton
+public class ExecutionFinalizer {
+    private static final Logger LOG = LogManager.getLogger(ExecutionFinalizer.class);
+
+    private final WorkflowDao workflowDao;
+    private final ExecutionDao executionDao;
+    private final GraphDao graphDao;
+
+    private final RenewableJwt internalUserCredentials;
+
+    private final Map<String, ManagedChannel> portalChannels;
+    private final LzyChannelManagerPrivateGrpc.LzyChannelManagerPrivateBlockingStub channelManagerClient;
+    private final GraphExecutorGrpc.GraphExecutorBlockingStub graphExecutorClient;
+    private final AllocatorGrpc.AllocatorBlockingStub allocatorClient;
+
+    public ExecutionFinalizer(WorkflowDao workflowDao, ExecutionDao executionDao, GraphDao graphDao,
+                              @Named("LzyServiceIamToken") RenewableJwt internalUserCredentials,
+                              @Named("PortalChannels") Map<String, ManagedChannel> portalChannels,
+                              @Named("ChannelManagerServiceChannel") ManagedChannel channelManagerChannel,
```

Comment:
If you create new type `PortalChannels` you may place all this channels there :)